File: src/ios/service/Service.swift
//
//  Service.swift
//
//  Created by Vikram Ezhil on 04/11/18.
//

import Foundation

protocol ServiceDelegate: class {
    //
    // Updates with the service running status
    //
    // :param: command The callback context
    // :param: passBackKey The pass back key
    // :param: serviceRunning The service running status
    //
    func onServiceRunning(command: CDVInvokedUrlCommand?, passBackKey: String, serviceRunning: Bool)
    
    //
    // Updates with the service response
    //
    // :param: command The callback context
    // :param: passBackKey The pass back key
    // :param: serviceResponse The service response
    //
    func onServiceResponse(command: CDVInvokedUrlCommand?, passBackKey: String, serviceResponse: String)
    
    //
    // Updates with the service error
    //
    // :param: command The callback context
    // :param: passBackKey The pass back key
    // :param: errorMessage The error message
    //
    func onServiceError(command: CDVInvokedUrlCommand?, passBackKey: String, errorMessage: String)
}

class Service: NSObject {
    private let TAG: String = "Service"

    weak var serviceDelegate: ServiceDelegate?
    
    ///
    /// Initializer
    ///
    override init() {
        super.init()
    }
    
    ///
    /// Sends a POST request
    ///
    /// :param: command The callback context
    /// :param: passBackKey The pass back key
    /// :param: requestLink The request link
    /// :param: customHeaders The custom headers if any
    /// :param: requestBody The request body
    ///
    /// :callbacks: ServiceDelegate onServiceRunning, onServiceResponse, onServiceError
    ///
    public func post(command: CDVInvokedUrlCommand?, passbackKey: String, requestLink: String, customHeaders: [String: String], requestBody: [String: Any]) {
        print(TAG, "Request type = POST");
        print(TAG, "Pass back key = \(passbackKey)");
        print(TAG, "Request link = \(requestLink)");
        print(TAG, "Custom headers = \(customHeaders)");
        print(TAG, "Request params = \(requestBody)");
        
        let request = NSMutableURLRequest(url: NSURL(string: requestLink)! as URL)
        request.httpMethod = "POST"
        request.addValue("application/json; charset=UTF-8", forHTTPHeaderField: "Content-Type")
        
        // Adding custom headers if applicable
        if customHeaders.count > 0 {
            request.allHTTPHeaderFields = customHeaders
        }
        
        // Setting the http body
        let jsonData = try? JSONSerialization.data(withJSONObject: requestBody, options: .prettyPrinted)
        request.httpBody = jsonData
        
        // Sending back the service running status
        serviceDelegate?.onServiceRunning(command: command, passBackKey: passbackKey, serviceRunning: true)

        let task = URLSession.shared.dataTask(with: request as URLRequest, completionHandler: {data, response, error -> Void in
            
            // Sending back the service running status
            self.serviceDelegate?.onServiceRunning(command: command, passBackKey: passbackKey, serviceRunning: false)
            
            if(error == nil) {
                if let httpResponse = response as? HTTPURLResponse {
                    if(httpResponse.statusCode == 200) {
                        if let responseData = data, let responseString = responseData.toString() {
                            self.serviceDelegate?.onServiceResponse(command: command, passBackKey: passbackKey, serviceResponse: responseString)
                        } else {
                            self.serviceDelegate?.onServiceError(command: command, passBackKey: passbackKey, errorMessage: "Data conversion error = \(String(describing: data))")
                        }
                    } else {
                        self.serviceDelegate?.onServiceError(command: command, passBackKey: passbackKey, errorMessage: "Service error, status code = \(httpResponse.statusCode)")
                    }
                } else {
                    self.serviceDelegate?.onServiceError(command: command, passBackKey: passbackKey, errorMessage: "Service error, status code unavailable")
                }
            } else {
                self.serviceDelegate?.onServiceError(command: command, passBackKey: passbackKey, errorMessage: error.debugDescription)
            }
        })
        
        task.resume()
    }
    
    
    ///
    /// Sends a GET request
    ///
    /// :param: command The callback context
    /// :param: passBackKey The pass back key
    /// :param: requestLink The request link
    /// :param: customHeaders The custom headers if any
    ///
    /// :callbacks: ServiceDelegate onServiceRunning, onServiceResponse, onServiceError
    ///
    public func get(command: CDVInvokedUrlCommand?, passbackKey: String, requestLink: String, customHeaders: [String: String]) {
        print(TAG, "Request type = GET");
        print(TAG, "Pass back key = \(passbackKey)");
        print(TAG, "Request link = \(requestLink)");
        print(TAG, "Custom headers = \(customHeaders)");
        
        let request = NSMutableURLRequest(url: NSURL(string: requestLink)! as URL)
        request.httpMethod = "GET"
        request.addValue("application/json; charset=UTF-8", forHTTPHeaderField: "Content-Type")
        
        // Adding custom headers if applicable
        if customHeaders.count > 0 {
            request.allHTTPHeaderFields = customHeaders
        }
        
        // Sending back the service running status
        serviceDelegate?.onServiceRunning(command: command, passBackKey: passbackKey, serviceRunning: true)
        
        let task = URLSession.shared.dataTask(with: request as URLRequest, completionHandler: {data, response, error -> Void in
            
            // Sending back the service running status
            self.serviceDelegate?.onServiceRunning(command: command, passBackKey: passbackKey, serviceRunning: false)
            
            if(error == nil) {
                if let httpResponse = response as? HTTPURLResponse {
                    if(httpResponse.statusCode == 200) {
                        if let responseData = data, let responseString = responseData.toString() {
                            self.serviceDelegate?.onServiceResponse(command: command, passBackKey: passbackKey, serviceResponse: responseString)
                        } else {
                            self.serviceDelegate?.onServiceError(command: command, passBackKey: passbackKey, errorMessage: "Data conversion error = \(String(describing: data))")
                        }
                    } else {
                        self.serviceDelegate?.onServiceError(command: command, passBackKey: passbackKey, errorMessage: "Service error, status code = \(httpResponse.statusCode)")
                    }
                } else {
                    self.serviceDelegate?.onServiceError(command: command, passBackKey: passbackKey, errorMessage: "Service error, status code unavailable")
                }
            } else {
                self.serviceDelegate?.onServiceError(command: command, passBackKey: passbackKey, errorMessage: error.debugDescription)
            }
        })
        
        task.resume()
    }
}

File: src/android/service/Service.java
package com.github.vikramezhil.cordova.service;

import android.content.Context;
import android.util.Log;

import com.android.volley.AuthFailureError;
import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.ServerError;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.HttpHeaderParser;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;

import org.apache.cordova.CallbackContext;

import org.json.JSONObject;

import java.util.Map;

/**
 * Service class
 *
 * @author Vikram Ezhil
 */

public class Service {
    private final String TAG = "Service";

    private Context context;
    private ServiceListener serviceListener;

    public interface ServiceListener {
        /**
         * Updates with the service running status
         *
         * @param callbackContext The callback context
         * @param passBackKey The pass back key
         * @param serviceRunning The service running status
         */
        void onServiceRunning(CallbackContext callbackContext, String passBackKey, Boolean serviceRunning);

        /**
         * Updates with the service response
         *
         * @param callbackContext The callback context
         * @param passBackKey The pass back key
         * @param serviceResponse The service response
         */
        void onServiceResponse(CallbackContext callbackContext, String passBackKey, String serviceResponse);

        /**
         * Updates with the service error
         *
         * @param callbackContext The callback context
         * @param passBackKey The pass back key
         * @param errorMessage The error message
         */
        void onServiceError(CallbackContext callbackContext, String passBackKey, String errorMessage);
    }

    /**
     * Service constructor
     *
     * @param context The application context
     * @param serviceListener The class instance which implements the service listener
     */
    public Service(Context context, ServiceListener serviceListener) {
        this.context = context;
        this.serviceListener = serviceListener;
    }

    /**
     * Sends a POST request
     *
     * @param callbackContext The callback context
     * @param passBackKey The pass back key
     * @param requestLink The request link
     * @param customHeaders The custom headers if any
     * @param requestBody The request body
     *
     * Listener Interface => ServiceListener
     */
    public void post(final CallbackContext callbackContext, final String passBackKey, final String requestLink, final Map<String, String> customHeaders, final JSONObject requestBody) {
        try {
            Log.wtf(TAG, "Request type = POST");
            Log.wtf(TAG, "Pass back key = " + passBackKey);
            Log.wtf(TAG, "Request link = " + requestLink);
            Log.wtf(TAG, "Request params = " + requestBody.toString());
            if(customHeaders != null) {
                Log.wtf(TAG, "Custom headers = " + customHeaders.toString());
            }

            serviceListener.onServiceRunning(callbackContext, passBackKey, true);

            RequestQueue requestQueue = Volley.newRequestQueue(context);
            StringRequest request = new StringRequest(Request.Method.POST, requestLink, new Response.Listener<String>() {
                @Override
                public void onResponse(String response) {
                    Log.wtf(TAG, "Response = " + response);

                    serviceListener.onServiceRunning(callbackContext, passBackKey, false);

                    serviceListener.onServiceResponse(callbackContext, passBackKey, response);
                }
            }, new Response.ErrorListener() {
                @Override
                public void onErrorResponse(VolleyError error) {
                    serviceListener.onServiceRunning(callbackContext, passBackKey, false);

                    NetworkResponse response = error.networkResponse;
                    if (error instanceof ServerError && response != null) {
                        try {
                            String res = new String(response.data, HttpHeaderParser.parseCharset(response.headers, "utf-8"));

                            serviceListener.onServiceError(callbackContext, passBackKey, res);

                            return;
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }

                    serviceListener.onServiceError(callbackContext, passBackKey, error.getLocalizedMessage());
                }
            }) {
                @Override
                public String getBodyContentType() {
                    return "application/json; charset=utf-8";
                }

                @Override
                public Map<String, String> getHeaders() throws AuthFailureError {
                    if(customHeaders != null) {
                        return customHeaders;
                    }

                    return super.getHeaders();
                }

                @Override
                public byte[] getBody() throws AuthFailureError {
                    try {
                        return requestBody.toString().getBytes("utf-8");
                    } catch (Exception e) {
                        e.printStackTrace();

                        Log.wtf(TAG, "Unsupported Encoding while trying to get the bytes of = " + requestBody.toString());

                        return null;
                    }
                }

                @Override
                protected Response<String> parseNetworkResponse(NetworkResponse response) {
                    return super.parseNetworkResponse(response);
                }
            };

            requestQueue.add(request);
        } catch (Exception e) {
            e.printStackTrace();

            serviceListener.onServiceRunning(callbackContext, passBackKey, false);
        }
    }

    /**
     * Sends a GET request
     *
     * @param callbackContext The callback context
     * @param passBackKey The pass back key
     * @param requestLink The request link
     * @param customHeaders The custom headers if any
     *
     * Listener Interface => ServiceListener
     */
    public void get(final CallbackContext callbackContext, final String passBackKey, final String requestLink, final Map<String, String> customHeaders) {
        try {
            Log.wtf(TAG, "Request type = GET");
            Log.wtf(TAG, "Pass back key = " + passBackKey);
            Log.wtf(TAG, "Request link = " + requestLink);
            if(customHeaders != null) {
                Log.wtf(TAG, "Custom headers = " + customHeaders.toString());
            }

            serviceListener.onServiceRunning(callbackContext, passBackKey, true);

            RequestQueue requestQueue = Volley.newRequestQueue(context);
            StringRequest request = new StringRequest(Request.Method.GET, requestLink, new Response.Listener<String>() {
                @Override
                public void onResponse(String response) {
                    Log.wtf(TAG, "Response = " + response);

                    serviceListener.onServiceRunning(callbackContext, passBackKey, false);

                    serviceListener.onServiceResponse(callbackContext, passBackKey, response);
                }
            }, new Response.ErrorListener() {
                @Override
                public void onErrorResponse(VolleyError error) {
                    serviceListener.onServiceRunning(callbackContext, passBackKey, false);

                    NetworkResponse response = error.networkResponse;
                    if (error instanceof ServerError && response != null) {
                        try {
                            String res = new String(response.data, HttpHeaderParser.parseCharset(response.headers, "utf-8"));

                            serviceListener.onServiceError(callbackContext, passBackKey, res);

                            return;
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }

                    serviceListener.onServiceError(callbackContext, passBackKey, error.getLocalizedMessage());
                }
            }) {
                @Override
                public String getBodyContentType() {
                    return "application/json; charset=utf-8";
                }

                @Override
                public Map<String, String> getHeaders() throws AuthFailureError {
                    if(customHeaders != null) {
                        return customHeaders;
                    }

                    return super.getHeaders();
                }

                @Override
                protected Response<String> parseNetworkResponse(NetworkResponse response) {
                    return super.parseNetworkResponse(response);
                }
            };

            requestQueue.add(request);
        } catch (Exception e) {
            e.printStackTrace();

            serviceListener.onServiceRunning(callbackContext, passBackKey, false);
        }
    }
}
